onnxruntime/core/optimizer/selectors_actions/actions.cc

```diff
@@ -0,0 +1,85 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "core/optimizer/selectors_actions/actions.h"
+#include "core/optimizer/selectors_actions/helpers.h"
+#include "core/optimizer/utils.h"
+
+using namespace ONNX_NAMESPACE;
+using namespace ::onnxruntime::common;
+namespace onnxruntime {
+
+namespace {
+
+// Check if a node involved in an optimization can be safely removed due to it only having outputs consumed by nodes
+// in the removal_set. If it has an output edge to a node outside of that set it must remain.
+// As we can't easily remove a NodeArg from the Node::OutputDefs for the node being removed, we do not check if the
+// node provides graph outputs here. The optimizer must correctly handle nodes producing graph outputs
+// and not attempt to delete one of those nodes unless it has created a new source for the graph output.
+bool CanSafelyRemoveNode(Node& node_to_remove, const std::unordered_set<const Node*>& removal_set) {
+  bool safe = true;
+  for (auto iter = node_to_remove.OutputEdgesBegin(), end = node_to_remove.OutputEdgesEnd(); iter != end; ++iter) {
+    if (removal_set.find(&iter->GetNode()) == removal_set.cend()) {
+      safe = false;
+      break;
+    }
+  }
+
+  return safe;
+}
+
+// remove nodes if it is 'safe' to do so according to the checks in CanSafelyRemoveNode.
+void SafelyRemoveNodes(Graph& graph, const std::vector<Node*>& nodes_to_remove, const Node* ignore_target) {
```

Comment:
>nodes_to_remove [](http://example.com/codeflow?start=63&length=15)

there is an assumption here: nodes_to_remove has no dup node or it will crash.

Reply:
There is, but this is just a local helper and only called via a NodesToOptimize instance. We could add checks but that feels like unnecessary code given the usage (and if it breaks it should break in testing the optimizer that is adding multiple instances of a nodes to the NodesToOptimize set).